EKF/control.cpp

```diff
@@ -977,6 +977,18 @@ void Ekf::controlMagFusion()
 				_control_status.flags.mag_hdg = true;
 			}
 
+			// perform switch-over from only updating the mag states to updating all states
+			if (!_control_status.flags.update_mag_states_only && _control_status_prev.flags.update_mag_states_only) {
+				// reset the mag field covariances
+				zeroRows(P, 16, 21);
+				zeroCols(P, 16, 21);
+
+				// re-instate the last used variances
+				for (uint8_t index = 0; index <= 5; index ++) {
+					P[index+16][index+16] = _saved_mag_variance[index];
```

Comment:
We do need to reset the off-diagonals (co-variances), but further thought is needed on how the diagonals (variances) should be set. The _saved_mag_variance value was intended to save the variances when switching in and out of 3-axis fusion, however in this case where we have remained in3-axis fusion the learned mag field states are valid, so there needs to be a good reason for why we would not continue to use them.